src/Microsoft.Repl.Tests/ShellTests.cs

```diff
@@ -159,19 +226,347 @@ public async Task RunAsync_WithUnhandledKeyPress_DoesNothing()
+        public async Task RunAsync_WithShiftTabKeyPressAndNoSuggestions_DoesNothing()
+        {
+            ConsoleKeyInfo consoleKeyInfo = new ConsoleKeyInfo(keyChar: '\0',
+                key: ConsoleKey.Tab,
+                shift: true,
+                alt: false,
+                control: false);
+            Shell shell = CreateShell(consoleKeyInfo,
+                caretPosition: 0,
+                previousCommand: null,
+                nextCommand: null,
+                out CancellationTokenSource cancellationTokenSource);
+
+            string inputBufferTextBeforeKeyPress = "z";
+
+            IShellState shellState = shell.ShellState;
+            shellState.InputManager.SetInput(shellState, inputBufferTextBeforeKeyPress);
+
+            await shell.RunAsync(cancellationTokenSource.Token).ConfigureAwait(false);
+
+            // Verify the input buffer contents after Shift + Tab key press event
+            Assert.Equal(inputBufferTextBeforeKeyPress, shell.ShellState.InputManager.GetCurrentBuffer());
+        }
+
+        [Fact]
+        public async Task RunAsync_WithEnterKeyPress_UpdatesInputBufferWithEmptyString()
+        {
+            ConsoleKeyInfo consoleKeyInfo = new ConsoleKeyInfo(keyChar: '\0',
+                key: ConsoleKey.Enter,
+                shift: false,
+                alt: false,
+                control: false);
+            Shell shell = CreateShell(consoleKeyInfo,
+                caretPosition: 0,
+                previousCommand: null,
+                nextCommand: null,
+                out CancellationTokenSource cancellationTokenSource);
+
+            string inputBufferContents = "set base \"https://localhost:44366/\"";
+
+            IShellState shellState = shell.ShellState;
+            shellState.InputManager.SetInput(shellState, inputBufferContents);
+
+            await shell.RunAsync(cancellationTokenSource.Token).ConfigureAwait(false);
+
+            Assert.Equal(string.Empty, shell.ShellState.InputManager.GetCurrentBuffer());
+        }
+
+        [Fact]
+        public async Task RunAsync_WithLeftArrowKeyPress_VerifyMoveCaretWasCalled()
+        {
+            ConsoleKeyInfo consoleKeyInfo = new ConsoleKeyInfo(keyChar: '\0',
+                key: ConsoleKey.LeftArrow,
+                shift: false,
+                alt: false,
+                control: false);
+            Shell shell = CreateShell(consoleKeyInfo,
+                caretPosition: 3,
+                previousCommand: null,
+                nextCommand: null,
+                out CancellationTokenSource cancellationTokenSource);
+
+            string inputBufferContents = "set base \"https://localhost:44366/\"";
+
+            IShellState shellState = shell.ShellState;
+            shellState.InputManager.SetInput(shellState, inputBufferContents);
+
+            await shell.RunAsync(cancellationTokenSource.Token).ConfigureAwait(false);
+
+            VerifyMoveCaretMethodWasCalled(shellState, -1, Times.Once());
```

Comment:
I dislike this validation approach.
1. It tests an implementation detail (was a method called a specified number of times with a given parameter).  This is inflexible to changes in the implementation that might set the caret position differently.  What if it were called with a different parameter? A different number of times? What if the implementation set the caret position directly instead of moving it by an offset?  Most importantly, should this test fail if any of these change?
2. It doesn't validate that the caret is in the expected position after the action is taken.

IMO the test should be asserting the caret position (the state of the system), not how it was moved.

Reply:
I am using a mock of IConsoleManager. So can't verify the caret position. This is the best I could come up with. Can we please discuss this offline when you have time? Thanks!

Reply:
@jimmylewis , as discussed offline, made changes here:
https://github.com/aspnet/HttpRepl/commit/f3eeb8fa792d5435e3712f073b06f1c0a3190f51

Could you please take a look and let me know what you think. Thanks!

Reply:
I like the approach, but there's some issues with it. Can you get that commit into this PR? (If I comment on the commit, they won't show up in any PR).

For example:
        public void MoveCaret(int position)
        {
            CaretPosition = position;
        }
`position` should be `offset`, and it should modify position but not overwrite it.  In the tests, you're asserting that the caret has a negative position - that doesn't make sense.  For example, when you start at position 3 and then go to the beginning, the test should assert that the caret position is now 0, not -3.

Reply:
Oops, sorry. Looks like I committed to wrong branch. Updated commit:
https://github.com/aspnet/HttpRepl/commit/cec25199e0d8fb991c1856832205ff381318af36